vertical-pod-autoscaler/initializer/initializer.go

```diff
@@ -0,0 +1,347 @@
+			}
+
+			container.Resources.Requests[resource] = recommended
+		}
+	}
+
+}
+
+// applyPolicy updates recommendation if recommended resources exceed limits defined in VPA resources policy
+func applyPolicy(recommendation *apimock.ContainerRecommendation, policy *apimock.ContainerPolicy) {
+	for resourceName, recommended := range recommendation.Resources {
+		if policy != nil {
+			resourcePolicy, found := policy.ResourcePolicy[resourceName]
+			if found {
+				if !resourcePolicy.Min.IsZero() && recommended.Value() < resourcePolicy.Min.Value() {
+					glog.Warningf("recommendation outside of policy bounds : min value : %v recommended : %v",
+						resourcePolicy.Min.Value(), recommended)
+					recommendation.Resources[resourceName] = resourcePolicy.Min
+				}
+				if !resourcePolicy.Max.IsZero() && recommended.Value() > resourcePolicy.Max.Value() {
+					glog.Warningf("recommendation outside of policy bounds : max value : %v recommended : %v",
+						resourcePolicy.Max.Value(), recommended)
+					recommendation.Resources[resourceName] = resourcePolicy.Max
+				}
+			}
+		}
+
+	}
+}
+
+func getRecommendationForContainer(recommendation *apimock.Recommendation, container v1.Container) *apimock.ContainerRecommendation {
+	for i, containerRec := range recommendation.Containers {
+		if containerRec.Name == container.Name {
+			return &recommendation.Containers[i]
+		}
+	}
+	return nil
+}
+
+func getContainerPolicy(containerName string, policy *apimock.ResourcesPolicy) *apimock.ContainerPolicy {
+	if policy != nil {
+		for _, container := range policy.Containers {
+			if containerName == container.Name {
+				return &container
+			}
+		}
+	}
+	return nil
+}
+
+// This should be cached as part of vpaLister.
+func (i *initializer) getMatchingVPA(pod *v1.Pod) *apimock.VerticalPodAutoscaler {
+	configs, err := i.vpaLister.List()
+	if err != nil {
+		//cannot get vpa configs
+		return nil
+	}
+	for _, vpaConfig := range configs {
+		selector, err := labels.Parse(vpaConfig.Spec.Target.Selector)
+		glog.V(4).Infof("VPA: :%+v", vpaConfig)
+		if err != nil {
+			continue
+		}
+		if selector.Matches(labels.Set(pod.GetLabels())) {
+			return vpaConfig
+		}
+		glog.V(4).Infof("Pod didn't match: %+v", pod.GetLabels())
+	}
+	return nil
+}
+
+// NewInitializer returns a VPA initializer.
+func NewInitializer(kubeClient kubeclient.Interface, cacheTTl time.Duration) Initializer {
+	i := &initializer{
+		client:      kubeClient,
+		vpaLister:   newVpaLister(kubeClient),
+		registerer:  newRegisterer(kubeClient),
+		recommender: recommender.NewCachingRecommender(cacheTTl, apimock.NewRecommenderAPI()),
+	}
+
+	informer := newInformer(kubeClient, i)
+	i.informer = informer
+	i.podListerSynced = i.informer.HasSynced
+
+	return i
+}
+
+func newVpaLister(kubeClient kubeclient.Interface) apimock.VerticalPodAutoscalerLister {
+	return apimock.NewVpaLister(kubeClient)
+}
+
+func newInformer(kubeClient kubeclient.Interface, i *initializer) cache.SharedInformer {
```

Comment:
you are passing a half-backed initiazliser, this is not a good patern. Please pass only the needed elements.